### PR TITLE
[Form] Refactor DefaultChoiceListFactory

### PR DESCRIPTION
**Pull request by vudaltsov**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

A small optimization (remove unnecessary `foreach` calls) and a naming improvement.

ping @HeahDude , @stof 

#EUFOSSA

### CODE REVIEW

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 1]
```diff
@@ -81,9 +81,21 @@ public function createView(ChoiceListInterface $list, $preferredChoices = null,
                     $otherViews
                 );
             }
+
+            foreach ($preferredViews as $key => $view) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: Can you keep the comment here please?

**Review by HeahDude — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php (hunk 1 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: @HeahDude , done!

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 2]
```diff
@@ -96,20 +110,6 @@ public function createView(ChoiceListInterface $list, $preferredChoices = null,
             );
         }
 
-        // Remove any empty group view that may have been created by
-        // addChoiceViewGroupedBy()
-        foreach ($preferredViews as $key => $view) {
```
Comment (webmozart) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: I don't think this change is backwards compatible. What if I create the following choices (which may come from a dynamic source):

```
$builder->add('field', ChoiceType::class, [
    'choices' => [
        'MyEmptyGroup' => [],
        'NonEmptyGroup' => [
            'A' => 'a_option',
            'B' => 'b_option',
        ],
    ],
]);
```

Before, `MyEmptGroup` would be removed. After, it wouldn't.

It's probably worth adding a test for this case.

**Review by webmozart — COMMENTED** [review]
Otherwise fine by me, although I don't see the benefit of changing those names.

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php (hunk 2 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: @webmozart , it is never created in this case, see https://github.com/symfony/symfony/blob/95d46c25f798519b51079d8360f85d615684b9f7/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php#L179

**Comment by vudaltsov** [comment]
Renaming makes the code easier to read. `addChoiceViewGroupedBy` and `addChoiceViewsGroupedBy` are almost impossible to distinguish.

**Review by webmozart — APPROVED** [review]
Well then :+1:

**Comment by fabpot** [comment]
Is it a bug fix or a refactoring? If that's the latter, it must be done on master.

**Comment by fabpot** [comment]
Thank you @vudaltsov.

**Pull request merged**